src/bin/metadata/mod.rs

```diff
@@ -0,0 +1,433 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use anchor_syn::idl::{Idl, IdlInstruction, IdlType, IdlTypeDefinitionTy};
+use clap::ArgMatches;
+use convert_case::{Boundary, Case, Casing};
+use serde_json::Value as JsonValue;
+use sha2::{Digest, Sha256};
+use solang_parser::lexer::KEYWORDS;
+use std::{
+    ffi::{OsStr, OsString},
+    fs::File,
+    io::Write,
+    path::PathBuf,
+    process::exit,
+};
+
+/// This subcommand generates a Solidity interface file from Anchor IDL file.
+/// The IDL file is json and lists all the instructions, events, structs, enums,
+/// etc. We have to avoid the numerous Solidity keywords, and retain any documentation.
+pub fn idl(matches: &ArgMatches) {
+    let files = matches.get_many::<OsString>("INPUT").unwrap();
+
+    let output = matches.get_one::<OsString>("OUTPUT").map(PathBuf::from);
+
+    for file in files {
+        idl_file(file, &output);
+    }
+}
+
+fn idl_file(file: &OsStr, output: &Option<PathBuf>) {
+    let f = match File::open(file) {
+        Ok(s) => s,
+        Err(e) => {
+            eprintln!("{}: error: {}", file.to_string_lossy(), e);
+            exit(1);
+        }
+    };
+
+    let idl: Idl = match serde_json::from_reader(f) {
+        Ok(idl) => idl,
+        Err(e) => {
+            eprintln!("{}: error: {}", file.to_string_lossy(), e);
+            exit(1);
+        }
+    };
+
+    let filename = format!("{}.sol", idl.name);
+
+    let path = if let Some(base) = output {
+        base.join(filename)
+    } else {
+        PathBuf::from(filename)
+    };
+
+    println!(
+        "{}: info: creating '{}'",
+        file.to_string_lossy(),
+        path.display()
+    );
+
+    let f = match File::create(&path) {
+        Ok(f) => f,
+        Err(e) => {
+            eprintln!("{}: error: {}", path.display(), e);
+            exit(1);
+        }
+    };
+
+    if let Err(e) = write_solidity(&idl, f) {
+        eprintln!("{}: error: {}", path.display(), e);
+        exit(1);
+    }
+}
+
+fn write_solidity(idl: &Idl, mut f: File) -> Result<(), std::io::Error> {
+    if let Some(program_id) = program_id(idl) {
+        writeln!(
+            f,
+            "anchor_{} constant {} = anchor_{}(address'{}');\n",
+            idl.name, idl.name, idl.name, program_id
+        )?;
+    }
+
+    let mut ty_names = idl
+        .types
+        .iter()
+        .map(|ty| (ty.name.to_string(), ty.name.to_string()))
+        .collect::<Vec<(String, String)>>();
+
+    if let Some(events) = &idl.events {
+        events
+            .iter()
+            .for_each(|event| ty_names.push((event.name.to_string(), event.name.to_string())));
+    }
+
+    rename_keywords(&mut ty_names);
+
+    for ty_def in &idl.types {
+        if let IdlTypeDefinitionTy::Enum { variants } = &ty_def.ty {
+            if variants.iter().any(|variant| variant.fields.is_some()) {
+                eprintln!(
+                    "enum {} has variants with fields, not supported in Solidity\n",
```

Comment:
While working on the NFT example, I have been thinking about unsupported types. Have you thought of transforming a `Option<Foo>` into `bool has_foo;` and `Foo foo;`?

Reply:
That might work for `Option<Foo>` in account storage, but when serialized in borsh the encoding is shorter for `None` (just a single 0 byte).